components/style/properties_and_values.rs

```diff
@@ -0,0 +1,1167 @@
+        //
+        // A 'specific ident' is any sequence consisting of a name-start code
+        // point, followed by zero
+        // or more name code points, which matches the <custom-ident>
+        // production
+        // https://drafts.css-houdini.org/css-properties-values-api-1/#supported-syntax-strings
+        //
+        // ...
+        // This generic data type is denoted by <custom-ident>, and represents
+        // any valid CSS identifier that would not be misinterpreted as a
+        // pre-defined keyword in that property’s value definition.
+        // https://drafts.csswg.org/css-values-4/#identifier-value
+        //
+        // So! In order to make sure specific identifiers don't contain
+        // escapes, we need to check for escapes, which are only introduced by
+        // backslashes, which shouldn't show up anyhow.
+        // https://drafts.csswg.org/css-syntax-3/#escaping
+        let mut contains_backslash = false;
+        for c in input.chars() {
+            if c == '\\' {
+                contains_backslash = true;
+                break
+            }
+        }
+        if contains_backslash {
+            return Err(())
+        }
+
+        // The parsed syntax string, which we'll build up as we scan tokens.
+        let mut syntax = None;
+
+        // The syntax string isn't really CSS, but hopefully this maximizes
+        // code reuse.
+        let mut parser_input = ParserInput::new(input);
+        let mut parser = Parser::new(&mut parser_input);
+
+        #[derive(Debug)]
+        enum State {
+            // *.
+            AfterAsterisk,
+            // <.
+            AfterOpenAngle,
+            // +.
+            AfterPlus,
+            // <type>.
+            // ident.
+            AfterType { after_whitespace: bool },
+            // <type.
+            AfterTypeName,
+            // .
+            // |.
+            Start { after_bar: bool },
+        }
+
+        let mut state = State::Start { after_bar: false };
+
+        /// Add a `Type` to the disjunction. It might turn out this is a list
+        /// term, in which case we'll modify the `Term` later.
+        fn push_type(syntax: &mut Option<Syntax>, t: Type) {
+            if let Some(Syntax::Disjunction(ref mut ts)) = *syntax {
+                ts.push(Term { typ: t, list: false })
+            } else { unreachable!() }
+        }
+
+        /// Signal that we expect to be parsing some term in a disjunction.
+        fn expect_disjunction(syntax: &mut Option<Syntax>) {
+            if let Some(Syntax::Disjunction(_)) = *syntax {
+                // Good!
+            } else {
+                assert!(*syntax == None);
+                *syntax = Some(Syntax::Disjunction(vec![]))
+            }
+        }
+
+        /// Handle the next token in the syntax string (including whitespace).
+        fn handle_token(syntax: &mut Option<Syntax>, state: State, token: &Token) -> Result<State, ()> {
```

Comment:
I _suspect_ that this can be cleaner having some sort of `SyntaxParser` object that contains the `Option<Syntax>`, and presumably the `State`, and implementing methods on it, instead of a lot of free functions. In rust in-out parameters aren't very common.

I suspect it's not a big deal, and can be cleaned up later, but it'd be really nice to see it. Your call.